subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCrossBuildInMemoryCacheFactory.java

```diff
@@ -235,7 +237,7 @@ protected V maybeGetRetainedValue(Class<?> key) {
         private Map<Class<?>, V> getCacheScope(Class<?> type) {
             ClassLoader classLoader = type.getClassLoader();
             if (classLoader instanceof VisitableURLClassLoader) {
-                return ((VisitableURLClassLoader) classLoader).getUserData(this, HashMap::new);
+                return ((VisitableURLClassLoader) classLoader).getUserData(this, ConcurrentHashMap::new);
```

Comment:
The `userData` field on `VisitableURLClassLoader` is not a concurrent hash map, but this method must be thread-safe as it is called from `retainValue`

Reply:
> The `userData` field on `VisitableURLClassLoader` is not a concurrent hash map, but this method must be thread-safe as it is called from `retainValue`

`getUserData` is synchronized though

Reply:
ahh missed that